build-logic/src/main/kotlin/kotlin-inject.jvm.gradle.kts

```diff
@@ -7,11 +7,16 @@ plugins {
 
 val libs = the<LibrariesForLibs>()
 
-kotlin.compilerOptions.jvmTarget = JvmTarget.JVM_11
+kotlin.compilerOptions.jvmTarget = libs.versions.jvmTarget.map(JvmTarget::fromTarget)
 
 java {
-    sourceCompatibility = JavaVersion.VERSION_11
-    targetCompatibility = JavaVersion.VERSION_11
+    toolchain {
+        languageVersion.set(libs.versions.jdk.map(JavaLanguageVersion::of))
+    }
```

Comment:
As per [Jake](https://kotlinlang.slack.com/archives/C3PQML5NU/p1689790193010619?thread_ts=1689789822.225749&cid=C3PQML5NU) regarding `toolchain`:

> A toolchain declaration will force the presence of JDK 17 rather than allowing any JDK 17 or newer to build your project. The Java compiler is designed to cross-compiler older releases. I would not recommend forcing old JDKs as there really should be no reason to use anything but the latest locally (or maybe N-1 while waiting for Gradle support).

Is setting Kotlin and detekt's jvmTarget not enough?

Reply:
I was going to link to https://jakewharton.com/gradle-toolchains-are-rarely-a-good-idea/ as well. I wonder if there's a way to just target this to detekt since that's where the actual issue is.

I've been running locally with java 21 and it's been working fine. I think that would be a better target than 17 if we need to use a specific jdk.

Reply:
Tbh I disagree with Jake! At least until Gradle (ever) adds support for a minimum jdk version. I find it eternally annoying to checkout gradle projects that have undefined tooling requirements and manifest with odd/surprising errors. I can limit this PR to just the jvmTarget in detekt if you really prefer though.

Reply:
What about checking the jdk version, and applying a toolchain if it's below the minimum?

Reply:
my [LOGAF](https://blog.danlew.net/2020/04/15/the-logaf-scale/) is not high enough to implement that for this PR :)

Reply:
> I find it eternally annoying to checkout gradle projects that have undefined tooling requirements and manifest with odd/surprising errors.

Fair enough, I think a toolchain set to the latest version of java that doesn't break everything would be an acceptable route

Reply:
sounds good 👍. Does that mean you want to raise this to 22?

Reply:
If 22 works then yep, would also recommend updating https://github.com/evant/kotlin-inject/blob/fe512838343f1234ebbdf87d7c73b94ea3bcad1c/.circleci/config.yml#L54